### ISSUE TITLE
Better DNF Autocompletions

### ISSUE THREAD

**Issue by exploide** [fish-shell/fish-shell#3062]
Nice to see the recent 2.3.0 release containing completions for the dnf package manager.

I noticed a few thing which could be improved and I want to collect them here.
- [ ] Completions do not occur immediately

When I open a new terminal window, type `dnf` and hit tab button, it tries to complete with available files and folders. After I delete the line I typed, retype `dnf` and hit tab button, the expected completions occur.
- [ ] Deprecated subcommand `update` is suggested

When trying to install updates, one would like to type `dnf up` and hit tab. Unfortunately, one has to select one out of `update` and `upgrade`. This is unnecessary because `update` is a deprecated alias of `upgrade` and should not be completed.
- [ ] Some subcommands are not in the list of subcommand completions, e.g. `repolist`, `copr`, `provides`, ...

I expect that there are more deprecated subcommands which should not be completed and some regular subcommands which are not in the list.

**Comment by werk-xyz** [comment]
>  Completions do not occur immediately
yep, also occurs on my fedora 23(x64) workstation

**Comment by faho** [comment]
Let me copy the entire dnf completion script:

```fish
complete -c dnf -w yum
```

Yeah. Most issues here are probably caused by it just being a wrapper around the yum completions (i.e. fish assumes `dnf` works _exactly_ like yum - which is _almost_ correct as we see here).

> Some subcommands are not in the list of subcommand completions, e.g. 

Are these exclusively in dnf?

>Deprecated subcommand update is suggested

Is this only removed in dnf or also in yum?

Also, I was under the impression that dnf was supposed to be renamed to yum once it became stable - has that plan been cancelled?

**Comment by exploide** [comment]
Ok, I checked the man pages
http://linux.die.net/man/8/yum
https://www.mankier.com/8/dnf

> Are these exclusively in dnf?

These are the commands available in yum:
```
check check-update clean deplist distribution-synchronization downgrade erase groupinfo groupinstall grouplist groupremove groupupdate help history info install list load-transaction localinstall localupdate makecache provides reinstall remove repolist resolvedep search shell update update-to upgrade upgrade-to version whatprovides 
```

These are the commands available in dnf (excluding the deprecated ones, they are not listed in the beginning of the man page):
```
autoremove check-update clean distro-sync downgrade group help history info install list makecache mark provides reinstall remove repolist repository-packages search updateinfo upgrade upgrade-to 
```

They look quite different. 

I parsed the command lists and made use of Python's set difference functionality.

The following commands are only available in yum:
```python
In [10]: yum.difference(dnf)
Out[10]: 
{'check',
 'deplist',
 'distribution-synchronization',
 'erase',
 'groupinfo',
 'groupinstall',
 'grouplist',
 'groupremove',
 'groupupdate',
 'load-transaction',
 'localinstall',
 'localupdate',
 'resolvedep',
 'shell',
 'update',
 'update-to',
 'version',
 'whatprovides'}
```

And the following commands are only available in DNF:

```python
In [11]: dnf.difference(yum)
Out[11]: 
{'autoremove',
 'distro-sync',
 'group',
 'mark',
 'repository-packages',
 'updateinfo'}
```

Additionally, there are plugins for DNF which are common for a Fedora installation, e.g. copr.

> Is this only removed in dnf or also in yum?

I could not find a source for that. It seems yum is going to die. According to my local `man yum`:

> YUM  executable  is  now powered by DNF.  If you are sure you really want to use YUM, it is still available as yum-deprecated.

I was also confused if dnf is going to be renamed to yum. I'm not really sure but for a while it looks more like dnf is the new name which is going to persist.

I guess for the long run, it is worth having a separate completion script for dnf since yum seems to be stalled for now and a separate completion script would allow to keep track of subsequent changes.

**Comment by jrobeson** [comment]
DNF isn't being renamed to yum at least during the next fedora release (24).

DNF will likely continue diverging. One of the main missing completions at this point is the copr plugin.

**Comment by infix** [comment]
I would like to work on this if no one else is working on it.

**Comment by faho** [comment]
@infix: Sure, why not? I'd recommend having a look at #4724.

**Comment by exploide** [comment]
Apparently fixed by 5abc129778d11d7d16ded0883e5bee29ee9a8b3c in #5948 
Thanks @infix
